### ISSUE TITLE
[Task 29] REINFORCE++ / REINFORCE++-baseline implementation proposal

### ISSUE THREAD

**Issue by howtomakeaname** [redai-infra/Relax#177]
Related to #86, Task 29.

## 0. Background & current state

The `reinforce_plus_plus` / `reinforce_plus_plus_baseline` advantage functions already exist in `ppo_utils.py`, but the two variants are **not registered in `ALGOS`**, so the Controller rejects the key (`Algorithm key 'reinforce_plus_plus' not registered in ALGOS`) and they cannot run end-to-end. This task makes them runnable and verifiable, and pins the formulas / normalization dimensions / mask / reduction semantics to resolve the "naming and baseline convention are not unified" gap.

Key findings (detailed in `docs/algorithms/reinforce_plus_plus.md`):

- `reinforce_plus_plus_baseline`'s group-mean baseline is subtracted upstream in `post_process_rewards` (**no std division**, unlike GRPO); the advantage function only broadcasts. This convention was scattered and undocumented.
- Advantage whitening (`--normalize-advantages`) is applied only in the sync path `loss.py:compute_advantages_and_returns`, not in the async path `components/advantages.py` (pre-existing behavior shared by all algorithms).

## 1. Repo / baseline / branch

- Repo: `redai-infra/Relax`
- Baseline commit: `05b2fb0` (main, latest remote main)
- Dev branch: `feature/reinforce-plus-plus`

## 2. Environment

- Image: `ghcr.io/redai-infra/relaxrl:latest` (Megatron-LM / SGLang / Ray preinstalled, version-matched); `pip install -e .` inside the container.
- Python 3.12, CUDA 12.9, torch 2.11, sglang 0.5.12, ray 2.56.1, Megatron-LM (bundled in the image).
- Hardware: NVIDIA A100-SXM4-80GB, sync colocate mode.

## 3. Launch scripts & commands

```bash
# Inside the official image (ghcr.io/redai-infra/relaxrl:latest),
# after cloning Relax and `pip install -e .`:
export MODEL_DIR=<path-to-Qwen3-0.6B> DATA_DIR=<path-to-data> CUDA_VISIBLE_DEVICES=0
bash examples/algorithms/run-qwen3-0.6B-1xgpu-reinforce-pp.sh          # REINFORCE++
bash examples/algorithms/run-qwen3-0.6B-1xgpu-reinforce-pp-baseline.sh # REINFORCE++-baseline
bash contributor-program/2026-cohort-1/run-qwen3-0.6B-1xgpu-grpo.sh   # GRPO baseline
```

New/changed params: `--advantage-estimator reinforce_plus_plus[_baseline]`, `--normalize-advantages`.

## 4. Files / functions / scope / non-goals

Changed files:

- `relax/core/registry.py`: register `reinforce_plus_plus` / `reinforce_plus_plus_baseline` in `ALGOS` (reuse GRPO topology, no critic); factor the duplicated GRPO-family topologies into a shared `_GRPO_TOPOLOGY` constant. **Key fix**: previously unregistered → could not run end-to-end.
- `tests/utils/training/test_ppo_utils_reinforce.py` (new): element-wise parity vs independent reference; variable-length, all-zero reward, single-sample, gamma discount, mask not polluting padding, fully-masked rejection; shared `compute_policy_loss` parity.
- `tests/backends/megatron/test_reinforce_pp_cp_parity.py` (new): CP gather/compute/slice wiring, baseline CP locality, DP masked-whiten invariance.
- `tests/core/test_registry_reinforce.py` (new): registration, topology reuse, no critic, `process_role` dispatch.
- `docs/algorithms/reinforce_plus_plus.md` (new): formulas / normalization / mask / reduction / variant comparison / CP·DP / known limitations / test coverage.
- `examples/algorithms/run-qwen3-0.6B-1xgpu-reinforce-pp.sh`, `...-reinforce-pp-baseline.sh` (new): colocate recipes.
- `relax/utils/training/ppo_utils.py`, `relax/utils/utils.py`: docstring/comment clarity on the baseline convention (**no behavior change**).

Non-goals:

- No `if algorithm == ...` branches (dispatch via the `ALGOS` registry).
- No change to default algorithm behavior or the PPO/critic path; the GRPO-family topology refactor is an equivalent dict, zero numerical regression.
- No whitening added to the async path (framework-wide async behavior, out of scope for this task).
- No `arguments.py` parsing changes (ask-first; reuse existing `--advantage-estimator` / `--normalize-advantages`).

## 5. Baseline data / acceptance metrics / statistics

- Data: GSM8K (`data/gsm8k/train.jsonl`, 7473 problems); AIME-2024 eval.
- Model: Qwen3-0.6B.
- Acceptance: (1) numerical — key tensors match an independent reference element-by-element (atol=1e-6); (2) tests — variable-length / all-zero reward / single-sample / DP·CP invariance all pass; (3) training — same-budget comparison of reward/loss/KL/throughput across algorithms, no NaN/Inf, stable.
- Statistics: same model/data/budget/hardware/batch/seq-len/sampling; ≥3 stable windows.
- Targets: all unit tests pass; training loss/reward sane; REINFORCE++ and baseline are behaviorally distinct and stable.

## 6. Delivery

GitHub Issue + Pull Request (code-type, per §1.2).
